src/Bicep.Decompiler/TemplateConverter.cs

```diff
@@ -655,7 +710,93 @@ StringSyntax createFakeModulePath(string templateLinkExpression)
+
+            // TODO implement when we have batchSize support (https://github.com/Azure/bicep/issues/1625)
+            TemplateHelpers.AssertUnsupportedProperty(copyProperty, "mode", "The \"mode\" property is not currently supported");
+            TemplateHelpers.AssertUnsupportedProperty(copyProperty, "batchSize", "The \"batchSize\" property is not currently supported");
+
+            return PerformScopedAction(() => {
+                resource = ExpressionHelpers.ReplaceFunctionExpressions(resource, function => {
+                    if (!StringComparer.OrdinalIgnoreCase.Equals(function.Function, "copyIndex"))
+                    {
+                        return;
+                    }
+    
+                    if (function.Parameters.Length == 0)
+                    {
+                        function.Function = "variables";
+                        function.Parameters = new LanguageExpression[]
+                        {
+                            new JTokenExpression(ResourceCopyLoopIndexVar),
+                        };
+                    }
+                    else if (function.Parameters.Length == 1)
+                    {
+                        if (function.Parameters[0] is JTokenExpression jTokenExpression && 
+                            jTokenExpression.Value.Type == JTokenType.String &&
+                            !ExpressionsEngine.IsLanguageExpression(jTokenExpression.Value.Value<string>()))
+                        {
+                            // leave it alone - named copy loops are used for property copies
+                            return;
+                        }
+
+                        var varExpression = new FunctionExpression(
+                            "variables",
+                            new LanguageExpression[]
+                            {
+                                new JTokenExpression(ResourceCopyLoopIndexVar),
+                            },
+                            Array.Empty<LanguageExpression>());
+
+                        function.Function = "add";
+                        function.Parameters = new LanguageExpression[]
+                        {
+                            varExpression,
+                            function.Parameters[0],
+                        };
+                    }
+                });
+
+                var bodySyntax = SyntaxFactory.CreateRangedForSyntax(ResourceCopyLoopIndexVar, ParseJToken(count), resourceBodyFunc(resource));
+                var decorators = new List<SyntaxBase>();
+                /* TODO implement when we have batchSize support (https://github.com/Azure/bicep/issues/1625)
+                if (mode is not null)
+                {
+                    decorators.Add(new DecoratorSyntax(
+                        SyntaxFactory.AtToken,
+                        SyntaxFactory.CreateFunctionCall("mode", new SyntaxBase[] {
```

Comment:
>CreateFunctionCall [](start = 38, length = 18)

The chosen syntax won't include the mode decorator anymore. It'll just be batchSize. (I don't mind uncommenting/fixing if this PR goes in before my batchSize one does.)

Reply:
Yeah I think easier to handle when it's merged - the decompilation tests check for error diagnostics on the baselines, so it'll be flagged if we miss this.